File: ByteTools.py
# Copyright (c) PROPHESSOR 2019-2020

import struct
from typing import BinaryIO

class ByteTools():
    def __init__(self, stream: BinaryIO):
        self.stream = stream
        self.order = 'little'

    def byteorder(self, order=None):
        if not order:
            return self.order

        self.order = order

    def seek(self, offset: int, mode='START'):
        whence = 0

        if mode == 'CUR': whence = 1

        self.stream.seek(offset, whence)

    def readASCIIString(self, length):
        return self.stream.read(length).decode('ascii')

    def readUnicodeString(self, length):
        return self.stream.read(length).decode('utf-8')

    def readString(self, length): return self.readUnicodeString(length)

    def getBytes(self, x):
        bytes = self.stream.read(x)

        if not bytes: raise IOError("Unexpected end of file")

        return bytes

    def readUInt(self, x):
        return int.from_bytes(self.getBytes(x), byteorder=self.order, signed=False)

    def readUInt8(self):  return self.readUInt(1)
    def readUInt16(self): return self.readUInt(2)
    def readUInt32(self): return self.readUInt(4)

    def readInt(self, x):
        return int.from_bytes(self.getBytes(x), byteorder=self.order, signed=True)

    def readInt8(self):  return self.readInt(1)
    def readInt16(self): return self.readInt(2)
    def readInt32(self): return self.readInt(4)
    def readInt64(self): return self.readInt(8)

    def readFloat(self):
        return struct.unpack('f', self.getBytes(4))

File: ByteTools.cpp
/**
 * ByteTools (C++) v1.1.1
 *
 * Copyright (c) PROPHESSOR 2019
 */

#pragma once

#include <iostream>
#include <cstdio>
#include <stdint.h>

#define SEEK_START SEEK_SET

class ByteTools {
    public:
        FILE    *m_buffer;

        ByteTools() {
            std::cout << "ByteTools (C++) v1.0.0 // Copyright (c) PROPHESSOR 2019" << std::endl;
        }

        ByteTools(const char* filename) {
            ByteTools();
            open(filename);
        }

        void open(const char* filename) {
            m_buffer = fopen(filename, "rb");
        }

        size_t tell(void) {
            return ftell(m_buffer);
        }

        void seek(long int offset, int mode) {
            fseek(m_buffer, offset, mode);
        }

        // read

        // int8

        int8_t readInt8() {
            int8_t tmp;

            fread(&tmp, sizeof(int8_t), 1, m_buffer);

            return tmp;
        }

        uint8_t readUInt8() {
            uint8_t tmp;

            fread(&tmp, sizeof(uint8_t), 1, m_buffer);

            return tmp;
        }

        // int16

        int16_t readInt16() {
            int16_t tmp;

            fread(&tmp, sizeof(int16_t), 1, m_buffer);

            return tmp;
        }

        uint16_t readUInt16() {
            uint16_t tmp;

            fread(&tmp, sizeof(uint16_t), 1, m_buffer);

            return tmp;
        }

        int16_t readInt16LE() {
            return readInt16();
        }

        uint16_t readUInt16LE() {
            return readUInt16();
        }

        // int32

        int32_t readInt32() {
            int32_t tmp;

            fread(&tmp, sizeof(int32_t), 1, m_buffer);

            return tmp;
        }

        uint32_t readUInt32() {
            uint32_t tmp;

            fread(&tmp, sizeof(uint32_t), 1, m_buffer);

            return tmp;
        }

        int32_t readInt32LE() {
            return readInt32();
        }

        uint32_t readUInt32LE() {
            return readUInt32();
        }
        
        // seek

        // int8

        int8_t seekInt8() {
            int8_t tmp;

            fread(&tmp, sizeof(int8_t), 1, m_buffer);
            seek(-sizeof(int8_t), SEEK_CUR);

            return tmp;
        }

        uint8_t seekUInt8() {
            uint8_t tmp;

            fread(&tmp, sizeof(uint8_t), 1, m_buffer);
            seek(-sizeof(uint8_t), SEEK_CUR);

            return tmp;
        }

        // int16

        int16_t seekInt16() {
            int16_t tmp;

            fread(&tmp, sizeof(int16_t), 1, m_buffer);
            seek(-sizeof(int16_t), SEEK_CUR);

            return tmp;
        }

        uint16_t seekUInt16() {
            uint16_t tmp;

            fread(&tmp, sizeof(uint16_t), 1, m_buffer);
            seek(-sizeof(uint16_t), SEEK_CUR);

            return tmp;
        }

        int16_t seekInt16LE() {
            return seekInt16();
        }

        uint16_t seekUInt16LE() {
            return seekUInt16();
        }

        // int32

        int32_t seekInt32() {
            int32_t tmp;

            fread(&tmp, sizeof(int32_t), 1, m_buffer);
            seek(-sizeof(int32_t), SEEK_CUR);

            return tmp;
        }

        uint32_t seekUInt32() {
            uint32_t tmp;

            fread(&tmp, sizeof(uint32_t), 1, m_buffer);
            seek(-sizeof(uint32_t), SEEK_CUR);

            return tmp;
        }

        int32_t seekInt32LE() {
            return seekInt32();
        }

        uint32_t seekUInt32LE() {
            return seekUInt32();
        }
        
        // skip

        // int8

        void skipInt8() {
            seek(sizeof(int8_t), SEEK_CUR);
        }

        void skipUInt8() {
            seek(sizeof(uint8_t), SEEK_CUR);
        }


        // int16

        void skipInt16() {
            seek(sizeof(int16_t), SEEK_CUR);
        }

        void skipUInt16() {
            seek(sizeof(uint16_t), SEEK_CUR);
        }

        void skipInt16LE() {
            seek(sizeof(int16_t), SEEK_CUR);
        }

        void skipUInt16LE() {
            seek(sizeof(uint16_t), SEEK_CUR);
        }

        // int32

        void skipInt32() {
            seek(sizeof(int32_t), SEEK_CUR);
        }

        void skipUInt32() {
            seek(sizeof(uint32_t), SEEK_CUR);
        }

        void skipInt32LE() {
            seek(sizeof(int32_t), SEEK_CUR);
        }

        void skipUInt32LE() {
            seek(sizeof(uint32_t), SEEK_CUR);
        }
};
